src/python/modules/ensembl/compara/plugins/TestFiles.py

```diff
@@ -0,0 +1,361 @@
+        target_path: Target's root path, e.g. "/home/user/pipelines/target".
+
+    Attributes:
+        ref_path (str): Reference's root path.
+        target_path (str): Target's root path.
+        common_tree (dict): Directory tree shared between reference and target paths.
+        ref_only (dict): Directory tree of directories/files only found in the reference.
+        target_only (dict): Directory tree of directories/files only found in the target.
+
+    Raises:
+        AssertionError: If reference or target paths do not exist.
+
+    """
+    def __init__(self, ref_path: str, target_path: str):
+        assert os.path.exists(ref_path), "Reference path '{}' not found".format(ref_path)
+        assert os.path.exists(target_path), "Target path '{}' not found".format(target_path)
+        # Get the directory trees for the given reference and target paths
+        self.ref_path = os.path.abspath(ref_path)
+        self._ref_tree_only = self._get_dir_cmp(self.ref_path)
+        if os.path.isfile(self.ref_path):
+            self.ref_path = os.path.dirname(self.ref_path)
+        self.target_path = os.path.abspath(target_path)
+        self._target_tree_only = self._get_dir_cmp(self.target_path)
+        if os.path.isfile(self.target_path):
+            self.target_path = os.path.dirname(self.target_path)
+        self.common_tree = copy.deepcopy(self.ref_only)
+        # Recursive nested function (closure) to compare reference and target trees and build the common one
+        def cmp_trees(ref_node: Dict, target_node: Dict, common_node: Dict) -> None:
+            """Compares reference and target trees, moving their shared structure to the common tree."""
+            ref_dirnames = set(ref_node.keys())
+            target_dirnames = set(target_node.keys())
+            # Remove directories only found on the reference tree
+            for dirname in ref_dirnames.difference(target_dirnames):
+                del common_node[dirname]
+            for dirname in common_node.keys():
+                if dirname != '.':
+                    cmp_trees(ref_node[dirname], target_node[dirname], common_node[dirname])
+                else:
+                    # Remove files only found on the reference tree
+                    for filename in ref_node[dirname].difference(target_node[dirname]):
+                        common_node[dirname].remove(filename)
+                    # Remove common files from reference and target trees
+                    for filename in common_node[dirname]:
+                        ref_node[dirname].remove(filename)
+                        target_node[dirname].remove(filename)
+                # Remove emptied directories
+                if not common_node[dirname]:
+                    del common_node[dirname]
+                if not ref_node[dirname]:
+                    del ref_node[dirname]
+                if not target_node[dirname]:
+                    del target_node[dirname]
```

Comment:
With this approach you only build the common content. My approach does also prune the reference and target trees at the same time. I did it this way because it is more efficient than your approach (less set algebra involved).